server/channels/app/post.go

```diff
@@ -773,51 +773,65 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
 		return false, nil
 	}
 
-	previewedPost, err := a.GetSinglePost(previewedPostID, false)
-	if err != nil {
-		if err.StatusCode == http.StatusNotFound {
+	previewedPost, appErr := a.GetSinglePost(previewedPostID, false)
+	if appErr != nil {
+		if appErr.StatusCode == http.StatusNotFound {
 			mlog.Warn("permalinked post not found", mlog.String("referenced_post_id", previewedPostID))
 			return false, nil
 		}
-		return false, err
-	}
-
-	channelMembers, err := a.GetChannelMembersPage(c, post.ChannelId, 0, 10000000)
-	if err != nil {
-		return false, err
+		return false, appErr
 	}
 
-	permalinkPreviewedChannel, err := a.GetChannel(c, previewedPost.ChannelId)
-	if err != nil {
-		if err.StatusCode == http.StatusNotFound {
+	permalinkPreviewedChannel, appErr := a.GetChannel(c, previewedPost.ChannelId)
+	if appErr != nil {
+		if appErr.StatusCode == http.StatusNotFound {
 			mlog.Warn("channel containing permalinked post not found", mlog.String("referenced_channel_id", previewedPost.ChannelId))
 			return false, nil
 		}
-		return false, err
+		return false, appErr
 	}
 
 	permalinkPreviewedPost := post.GetPreviewPost()
-	for _, cm := range channelMembers {
-		if permalinkPreviewedPost != nil {
-			post.Metadata.Embeds[0].Data = permalinkPreviewedPost
-		}
+	if permalinkPreviewedPost != nil {
+		post.Metadata.Embeds[0].Data = permalinkPreviewedPost
+	} else {
+		return false, nil
+	}
 
-		postForUser := a.sanitizePostMetadataForUserAndChannel(c, post, permalinkPreviewedPost, permalinkPreviewedChannel, cm.UserId)
+	// This is a hack to cache the JSON representation of the post metadata
+	// to be attached to the websocket event if the user doesn't have permission.
+	// We wipe the data, get the JSON, and restore the data again.
+	// This simplifies the data handling in the channel hook, where otherwise
+	// multiple goroutines would be racing to modify the Metadata field.
+	// And deep-copying the Embed data also has challenges because the type is "any".
+	post.Metadata.Embeds[0].Data = nil
+	modifiedJSON, err := post.ToJSON()
+	if err != nil {
+		mlog.Warn("Failed to encode post to JSON", mlog.Err(err))
+	}
+	post.Metadata.Embeds[0].Data = permalinkPreviewedPost
 
-		// Using DeepCopy here to avoid a race condition
-		// between publishing the event and setting the "post" data value below.
-		messageCopy := message.DeepCopy()
-		broadcastCopy := messageCopy.GetBroadcast()
-		broadcastCopy.UserId = cm.UserId
-		messageCopy.SetBroadcast(broadcastCopy)
+	// Modifying post for the original sender, in case they don't have permission.
+	// This wipes off the original post content. And then the marshalled post content
+	// in the websocket event is taken care of in the channel hook.
+	if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, post.UserId, permalinkPreviewedChannel) {
+		post.Metadata.Embeds[0].Data = nil
```

Comment:
I'm still confused here: we set above, then unset, then set again, and now we unset if the posting user doesn't have permission. Where do we set it for users that /should/ get it?

Reply:
Via the websocket event :) That's what I tried to clarify in the comment. But it's hard to grasp.

Essentially, one object is returned from the REST API endpoint, the other objects are sent via websocket. L818 is about the object that gets returned from the REST endpoint. For everyone else (including the posting user), it gets taken care of by the websocket event.

Reply:
I wanted to test this behaviour to see what's actually happening. I made a private channel in which the left user was a member, but the right user was not. On master, either user posting a permalink to a post in that channel correctly displays for the left user and not for the right user. 

Then I checked out this PR's branch, refreshed both clients, and posted the same permalink again for each user. Now, neither user sees the permalink resolve, even though the left user should have access to the channel.

Can you reproduce this locally, @agnivade?

![CleanShot 2023-06-23 at 10 00 46@2x](https://github.com/mattermost/mattermost/assets/1023171/f1170ae8-876c-44ca-8143-050971068f63)